lib/elastic_apm/middleware.rb

```diff
@@ -56,13 +56,29 @@ def start_transaction(env)
     end
 
     def trace_context(env)
-      return unless (header = env['HTTP_ELASTIC_APM_TRACEPARENT'])
-      TraceContext.parse(header)
+      return unless (header = get_traceparent_header(env))
+
+      TraceContext.parse(header).tap do |tc|
+        next unless (header = env['HTTP_TRACESTATE'])
+        tc.tracestate = TraceContext::Tracestate.parse(header)
+      end
     rescue TraceContext::InvalidTraceparentHeader
```

Comment:
What about errors that might be raised from parsing the `tracestate` header? Those are just raised?

I also find it confusing that there is a `TraceContext::Tracestate` object but calling `parse` on `TraceContext` parses the `traceparent` headers. Does the term "Trace Context" only apply to `traceparent` headers? If not, it would make more sense to have a `TraceContext::TraceParent` object.

In any case, there are two `parse` calls in this method: `TraceContext.parse(header)` . and `TraceContext::Tracestate.parse(header)`. Because only the first one can raise the `TraceContext::InvalidTraceparentHeader` error, it makes more sense to me to have the flow be:

parse the traceparent: `TraceContext.parse(header)`
catch the error: `rescue TraceContext::InvalidTraceparentHeader`
parse the tracestate: `TraceContext::Tracestate.parse(header)`

But maybe we should define an error that can be raised by parsing the `tracestate` and we should catch it.

Reply:
I see your point 👍

Reply:
If you just think of `TraceContext` as the `Traceparent` object, it makes more sense?

Reply:
Re: raising. `InvalidTraceparentHeader` is an exception we raise ourselves if the header doesn't match the strictly defined shape. I don't think we need strict parsing for the `Tracestate` header as it's more loosely and the only thing we do with it (for now) is passing it on as is.